core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1178,8 +1178,13 @@ private[spark] class DAGScheduler(
     listenerBus.post(SparkListenerTaskStart(task.stageId, stageAttemptId, taskInfo))
   }
 
-  private[scheduler] def handleSpeculativeTaskSubmitted(task: Task[_]): Unit = {
-    listenerBus.post(SparkListenerSpeculativeTaskSubmitted(task.stageId, task.stageAttemptId))
+  private[scheduler] def handleSpeculativeTaskSubmitted(task: Task[_], taskIndex: Int): Unit = {
+    val speculativeTaskSubmittedEvent =
+      SparkListenerSpeculativeTaskSubmitted(task.stageId, task.stageAttemptId)
+    // add taskIndex field for Executor Dynamic Allocation
+    speculativeTaskSubmittedEvent.updateTaskIndex(taskIndex)
+    speculativeTaskSubmittedEvent.updatePartitionId(task.partitionId)
+    listenerBus.post(speculativeTaskSubmittedEvent)
```

Comment:
How about adding a new apply() to `SparkListenerSpeculativeTaskSubmitted` to simplify the construction? e.g.,
```scala
object SparkListenerSpeculativeTaskSubmitted {
  def apply(stageId: Int, stageAttemptId: Int, taskIndex: Int, partitionId: Int)
    : SparkListenerSpeculativeTaskSubmitted = {
    val speculativeTaskSubmittedEvent =
      SparkListenerSpeculativeTaskSubmitted(stageId, stageAttemptId)
    // add taskIndex field for Executor Dynamic Allocation
    speculativeTaskSubmittedEvent.updateTaskIndex(taskIndex)
    speculativeTaskSubmittedEvent.updatePartitionId(task.partitionId)
    speculativeTaskSubmittedEvent
  }
```
so here we can construct the `SparkListenerSpeculativeTaskSubmitted` like `      SparkListenerSpeculativeTaskSubmitted(task.stageId, task.stageAttemptId, taskIndex, task.partitionId)
`

Reply:
How about this:
```
case class SparkListenerSpeculativeTaskSubmitted(
    stageId: Int,
    stageAttemptId: Int = 0)
  extends SparkListenerEvent {
  // Note: this is here for backwards-compatibility with older versions of this event which
  // didn't stored taskIndex
  private var _taskIndex: Int = -1
  private var _partitionId: Int = -1

  def taskIndex: Int = _taskIndex
  def partitionId: Int = _partitionId

  def this(stageId: Int, stageAttemptId: Int, taskIndex: Int, partitionId: Int) = {
    this(stageId, stageAttemptId)
    _partitionId = partitionId
    _taskIndex = taskIndex
  }
}
```

We can constrict with `taskIndex` and `partitionId`.
```
val speculativeTaskSubmittedEvent = new SparkListenerSpeculativeTaskSubmitted(
      task.stageId, task.stageAttemptId, taskIndex, task.partitionId)
```

default usage as
```
val speculativeTaskSubmittedEvent  = speculativeTaskSubmittedEvent(stageId, stageAttempId)
```

Reply:
Sounds better.